lib/Controller/UsersController.php

```diff
@@ -469,6 +501,254 @@ public function create($username, $password, array $groups= [], $email='') {
 		);
 	}
 
+	/**
+	 * Set password for user using link
+	 *
+	 * @PublicPage
+	 * @NoCSRFRequired
+	 * @NoAdminRequired
+	 * @NoSubadminRequired
+	 *
+	 * @param string $token
+	 * @param string $userId
+	 * @return TemplateResponse
+	 */
+	public function setPasswordForm($token, $userId) {
+		try {
+			$this->checkPasswordSetToken($token, $userId);
+		} catch (UserTokenException $e) {
+			if ($e->getCode() === UserTokenException::TOKEN_TIME_EXPIRED) {
+				return new TemplateResponse(
+					'user_management', 'new_user/resendtokenbymail',
+					[
+						'link' => $this->urlGenerator->linkToRouteAbsolute('user_management.Users.resendToken', ['userId' => $userId])
+					], 'guest'
+				);
+			}
+			$this->log->logException($e, ['app' => 'user_management']);
+			$tokenMessage = null;
+			if ($e->getCode() === UserTokenException::TOKEN_AND_TIME_INVALID) {
+				$tokenMessage = $this->l10n->t('The token provided is invalid.');
+			} else {
+				$tokenMessage = $this->l10n->t('The token mismatch happened.');
+			}
+			return new TemplateResponse(
+				'core', 'error',
+				[
+					"errors" => [["error" => $tokenMessage]]
+				], 'guest'
+			);
+		}
+
+		return new TemplateResponse(
+			'user_management', 'new_user/setpassword',
+			[
+				'link' => $this->urlGenerator->linkToRouteAbsolute('user_management.Users.setPassword', ['userId' => $userId, 'token' => $token])
+			], 'guest'
+		);
+	}
+
+	/**
+	 * @param string $token
+	 * @param string $userId
+	 * @return null
+	 * @throws UserTokenException
+	 */
+	private function checkPasswordSetToken($token, $userId) {
+		$user = $this->userManager->get($userId);
+
+		$splittedToken = \explode(':', $this->config->getUserValue($userId, 'owncloud', 'lostpassword', null));
+		if (\count($splittedToken) !== 2) {
+			$this->config->deleteUserValue($userId, 'owncloud', 'lostpassword');
+			throw new UserTokenException(UserTokenException::TOKEN_AND_TIME_INVALID);
+		}
+
+		if ($splittedToken[0] < ($this->timeFactory->getTime() - 60*60*12) ||
```

Comment:
make this expiration time an oc_appconfig value, even if undocumented. better than hard-coded

Reply:
Done